### PR TITLE
Error on nan inf

### PR DESCRIPTION
**Pull request by davidw**
The C code used to connect a C node to Erlang should explicitly disallow floats that Erlang does not accept.

It's better to fail early and in a clear way than have Erlang spit out something that the C programmer may very well not understand.

http://erlang.org/pipermail/erlang-questions/2014-December/082003.html

### CODE REVIEW

**Comment by OTP-Maintainer** [comment]
./otp_build all -a failed:

    orkspace/otp/lib/erl_interface/bin/x86_64-unknown-linux-gnu/erl_call
    /ldisk/jenkins/jobs/pullrequests/workspace/otp/lib/erl_interface/obj/x86_64-unknown-linux-gnu/libei.a(encode_double.o): In function `ei_encode_double':
    /ldisk/jenkins/jobs/pullrequests/workspace/otp/lib/erl_interface/src/encode/encode_double.c:34: undefined reference to `__fpclassify'
    collect2: error: ld returned 1 exit status
    make[4]: *** [/ldisk/jenkins/jobs/pullrequests/workspace/otp/lib/erl_interface/bin/x86_64-unknown-linux-gnu/erl_call] Error 1
    make[4]: Leaving directory `/ldisk/jenkins/jobs/pullrequests/workspace/otp/lib/erl_interface/src'
    make[3]: *** [opt] Error 2
    make[3]: Leaving directory `/ldisk/jenkins/jobs/pullrequests/workspace/otp/lib/erl_interface/src'
    make[2]: *** [opt] Error 2
    make[2]: Leaving directory `/ldisk/jenkins/jobs/pullrequests/workspace/otp/lib/erl_interface'
    make[1]: *** [opt] Error 2
    make[1]: Leaving directory `/ldisk/jenkins/jobs/pullrequests/workspace/otp/lib'
    make: *** [libs] Error 2



----------------

I am a script, I am not human 

----------------

**Comment by davidw** [comment]
Right, it needs to link to the math library, but I'm not sure how to accomplish that in all the build stuff.

**Comment by tuncer** [comment]
You'll have to modify `erl_interface/configure.in` to let autoconf find the lib `fpclassify` is implemented in. For inspiration you can take a look at uses of `AC_CHECK_FUNCS` or `AC_SEARCH_LIBS` in `erts/configure.in`. Also, anytime you modify `configure.in`, don't forget to run `./otp_build autoconf` from the top dir for regenerating the `configure` scripts.

**Comment by nox** [comment]
The rest of the VM just uses isfinite() and fallbacks on finite() if unavailable, you should probably use that.

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by psyeugenic** [comment]
Much better solution to use isfinite than fpclassify.

As an added bonus, erl_interface has to work on VxWorks. Hence my reluctance to change things in erl_interface. I'm unsure (atm) if all VxWorks platforms has math support. They certainly don't have float support all the time. If we don't find a suitable isfinite in configure what should happen? We can't just bail out.

**Comment by davidw** [comment]
If it has no float support, could it even send Inf or NaN?  The point of this change, for me, is to have those error out as close as possible to where they originate.

**Comment by psyeugenic** [comment]
Yes, fail faster is a good motto. My concern is if during buildtime we try to link to a library that does not exists. libm does not exist on VxWorks hence the problem.

I think the following should be done: Configure should check for the capability of isfinite (and libm existence) like now but use it with definition `HAVE_ISFINITE` (or what it is called) in `config.h.in` . Include `config.h` and have compile time condition around the usage of isfinite in encode_double and erl_term to return the error if does not have the capability.

File: lib/erl_interface/configure.in [hunk 1]
```diff
@@ -138,6 +138,7 @@ esac
 # Checks for libraries.
 AC_CHECK_LIB([nsl], [gethostbyname])
 AC_CHECK_LIB([socket], [getpeername])
+AC_CHECK_LIB([m], [isfinite])
```
Comment (nox) [inline lib/erl_interface/configure.in]: The existence of isfinite (and fallback on finite) is done differently in the rest of OTP.

**Comment by davidw** [comment]
So perhaps copying this:

https://github.com/erlang/otp/blob/maint/erts/configure.in#L2115

Is what we need?

**Comment by psyeugenic** [comment]
I think you need lines https://github.com/erlang/otp/blob/maint/erts/configure.in#L2115-L2125 but yes.

**Comment by OTP-Maintainer** [comment]
The summary line of the commit message is too long and/or ends with a "."
Make sure the whole message follows the guidelines here: https://github.com/erlang/otp/wiki/Writing-good-commit-messages.

Bad message: Copy configure.in rules, only run isfinite check if it's available


----------------

I am a script, I am not human 

----------------

**Comment by psyeugenic** [comment]
One final thing: You should use `#if defined(HAVE_ISFINITE)` instead of `#if HAVE_ISFINITE == 1`
Otherwise it looks good. =)

**Comment by OTP-Maintainer** [comment]
The summary line of the commit message is too long and/or ends with a "."
Make sure the whole message follows the guidelines here: https://github.com/erlang/otp/wiki/Writing-good-commit-messages.

Bad message: Copy configure.in rules, only run isfinite check if it's available


----------------

I am a script, I am not human 

----------------

**Comment by OTP-Maintainer** [comment]
The summary line of the commit message is too long and/or ends with a "."
Make sure the whole message follows the guidelines here: https://github.com/erlang/otp/wiki/Writing-good-commit-messages.

Bad message: Copy configure.in rules, only run isfinite check if it's available


----------------

I am a script, I am not human 

----------------

**Comment by OTP-Maintainer** [comment]
The summary line of the commit message is too long and/or ends with a "."
Make sure the whole message follows the guidelines here: https://github.com/erlang/otp/wiki/Writing-good-commit-messages.

Bad message: Copy configure.in rules, only run isfinite check if it's available


----------------

I am a script, I am not human 

----------------

**Comment by psyeugenic** [comment]
merged to master in some modified form

**Comment by davidw** [comment]
Cool, thanks!

**Pull request closed without merging**